src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SuppressGCTransitionAttribute.cs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// An attribute used to indicate a GC transition should be skipped when making an unmanaged function call.
+    /// </summary>
+    /// <example>
+    /// Example of a valid use case. The Win32 `GetTickCount()` function is a small performance related function
+    /// that reads some global memory and returns the value. In this case, the GC transition overhead is significantly
+    /// more than the memory read.
+    /// <code>
+    /// using System;
+    /// using System.Runtime.InteropServices;
+    /// class Program
+    /// {
+    ///     [DllImport("Kernel32")]
+    ///     [SuppressGCTransition]
+    ///     static extern int GetTickCount();
+    ///     static void Main()
+    ///     {
+    ///         Console.WriteLine($"{GetTickCount()}");
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    /// <remarks>
+    /// This attribute is ignored if applied to a method without the <see cref="System.Runtime.InteropServices.DllImportAttribute"/>.
+    ///
+    /// Supressing the GC transition is an advanced scenario and should not be done without fully understanding
+    /// potential consequences. Examples of cases where this attribute is not appropriate:
+    ///   * Native function calls back into the runtime (e.g. Reverse P/Invoke).
+    ///   * Native function performs some type of I/O.
+    ///   * Native function executes for a non-trivial amount of time.
+    ///   * Native function requires extensive object manipulation during marshalling.
+    ///
+    ///  Cases where the attribute can be considered:
```

Comment:
I think this should be worded in more prescriptive way. Right now, this is worded as examples. It leaves a lot of gray area, with a lot of reference to runtime implementation details. Something like:

P/Invoke method that this attribute is applies must have all of the following properies:
- Native function always executes for a trivial amount of time (less than 1 microsecond).
- Native function does not performs syscall (e.g. any type of I/O).
- Native function does not call back into the runtime (e.g. Reverse P/Invoke).
- Native function does not throw exceptions
- Native function does not take locks

The benefits of applying this attribute to P/Invoke method is reduction of the P/Invoke overhead by approximately 20 processor cycles (fill in the actual number).

Consequences of invalid uses of this attribute:
...

Reply:
Given these restrictions, it seems like the number of places this could be used on native functions one doesn't completely control is severely limited, e.g. I have to assume this can't be used with pretty much any function exposed from kernel32.dll. Is that the right way to think about this, that it's really intended for one's own native functions being called out to? GetTickCount was used as an example... as a code reviewer, how do I know that's safe?

Reply:
You have to know how the OS is implemented.

Reply:
@stephentoub I don't think there's an easy way to say it is safe. I'd say it is not safe unless proven otherwise. Some operating system APIs related to time are almost guaranteed to be implemented in certain way (eg. reading a value from shared page) and it is documented. Any usage in CoreCLR/CoreFX should likely also be backed with a benchmark.

Anything that is an internal call wrapper around P/Invoke today should automatically qualify for this optimization.

Reply:
Also, I expect a primary usage outside of CoreFX/CoreCLR to be in gaming (ie. Unity3D) where they had a requirement for this and in Mono they were implementing it in using custom internal call. Similarly, runtime bridges (eg. Xamarin.Mac) may use this to call some performance critical code for reference counting.

Reply:
Is there some stress mode we are currently exposing or could expose that users could run to help validate cases?

Even with documentation/etc, it would be hard to know that `QueryPerformanceCounter`/`QueryPerformanceFrequency` (for example) aren't doing anything that would negatively impact things. -- For example, on modern CPUs this is just a read of a model-specific register (and some code to handle power mode transitions, etc); but on older CPUs it is a query into the High Performance Eventing Timer (but is still designed to execute fairly quickly).

Reply:
@tannergooding As in something akin to GCStress? I think that is an interesting idea. I don't know enough about the feature to know if it would be able to indicate degenerate use cases though. Anyone have an idea on what a new one would look like?

Reply:
> As in something akin to GCStress?

Exactly.

Certain GCStress modes can already be beneficial for catching places where you are doing invalid operations.

For example, various operations around `Unsafe.AsPointer` that people try to do arent valid, because a GC collection "could" happen on the ref.

The downside is that GCStress nodes are only available in debug/checked builds.

If a similar mode existed for this attribute (or if an existing mode already covers it), that may be beneficial.

Reply:
I think that the most likely incorrect use of this attribute is native method that takes too long and negatively affects GC pause times. GCStress is not going to help with catching incorrect use like this.

Reply:
> I think that the most likely incorrect use of this attribute is native method that takes too long and negatively affects GC pause times.

Would it be possible to have a mode or tool which measures the time of these functions specifically to help identify potentially problematic cases? Obviously not for use in production, but as a mechanism to help users identify cases that are clearly incorrect.

Or, would you consider this something where users following documentation and profiling individual methods for call time would be sufficient?